// framer-motion@11.12.0/dist/es/motion/index.js
import { jsxs as M, jsx as y } from "react/jsx-runtime";
import { forwardRef as E, useContext as f } from "react";
import { MotionConfigContext as x } from "../context/MotionConfigContext.js";
import { MotionContext as C } from "../context/MotionContext/index.js";
import { useVisualElement as g } from "./utils/use-visual-element.js";
import { useMotionRef as L } from "./utils/use-motion-ref.js";
import { useCreateMotionContext as j } from "../context/MotionContext/create.js";
import { loadFeatures as P } from "./features/load-features.js";
import { isBrowser as w } from "../utils/is-browser.js";
import { LayoutGroupContext as S } from "../context/LayoutGroupContext.js";
import { LazyContext as b } from "../context/LazyContext.js";
import { motionComponentSymbol as h } from "./utils/symbol.js";
import { warning as N, invariant as R } from "../utils/errors.js";
import { featureDefinitions as k } from "./features/definitions.js";
function U({ preloadedFeatures: o, createVisualElement: t, useRender: n, useVisualState: e, Component: a }) {
  o && P(o);
  function v(r, p) {
    let m;
    const u = {
      ...f(x),
      ...r,
      layoutId: z(r)
    }, { isStatic: s } = u, i = j(r), c = e(r, s);
    if (!s && w) {
      D(u, o);
      const d = G(u);
      m = d.MeasureLayout, i.visualElement = g(a, c, u, t, d.ProjectionNode);
    }
    return M(C.Provider, { value: i, children: [m && i.visualElement ? y(m, { visualElement: i.visualElement, ...u }) : null, n(a, r, L(c, i.visualElement, p), c, s, i.visualElement)] });
  }
  const l = E(v);
  return l[h] = a, l;
}
function z({ layoutId: o }) {
  const t = f(S).id;
  return t && o !== void 0 ? t + "-" + o : o;
}
function D(o, t) {
  const n = f(b).strict;
  if (process.env.NODE_ENV !== "production" && t && n) {
    const e = "You have rendered a `motion` component within a `LazyMotion` component. This will break tree shaking. Import and render a `m` component instead.";
    o.ignoreStrict ? N(!1, e) : R(!1, e);
  }
}
function G(o) {
  const { drag: t, layout: n } = k;
  if (!t && !n)
    return {};
  const e = { ...t, ...n };
  return {
    MeasureLayout: t != null && t.isEnabled(o) || n != null && n.isEnabled(o) ? e.MeasureLayout : void 0,
    ProjectionNode: e.ProjectionNode
  };
}
export {
  U as createRendererMotionComponent
};
