// framer-motion@11.12.0/dist/es/components/AnimatePresence/PresenceChild.js
import { jsx as a } from "react/jsx-runtime";
import * as C from "react";
import { useId as h, useCallback as M, useMemo as m } from "react";
import { PresenceContext as v } from "../../context/PresenceContext.js";
import { useConstant as w } from "../../utils/use-constant.js";
import { PopChild as x } from "./PopChild.js";
const L = ({ children: n, initial: c, isPresent: r, onExitComplete: o, custom: u, presenceAffectsLayout: i, mode: p }) => {
  const e = w(z), l = h(), s = M((t) => {
    e.set(t, !0);
    for (const f of e.values())
      if (!f)
        return;
    o && o();
  }, [e, o]), d = m(
    () => ({
      id: l,
      initial: c,
      isPresent: r,
      custom: u,
      onExitComplete: s,
      register: (t) => (e.set(t, !1), () => e.delete(t))
    }),
    /**
     * If the presence of a child affects the layout of the components around it,
     * we want to make a new context value to ensure they get re-rendered
     * so they can detect that layout change.
     */
    i ? [Math.random(), s] : [r, s]
  );
  return m(() => {
    e.forEach((t, f) => e.set(f, !1));
  }, [r]), C.useEffect(() => {
    !r && !e.size && o && o();
  }, [r]), p === "popLayout" && (n = a(x, { isPresent: r, children: n })), a(v.Provider, { value: d, children: n });
};
function z() {
  return /* @__PURE__ */ new Map();
}
export {
  L as PresenceChild
};
