// framer-motion@11.12.0/dist/es/frameloop/batcher.js
import { createRenderStep as E } from "./render-step.js";
const a = [
  "read",
  // Read
  "resolveKeyframes",
  // Write/Read/Write/Read
  "update",
  // Compute
  "preRender",
  // Compute
  "render",
  // Write
  "postRender"
  // Compute
], F = 40;
function B(o, l) {
  let t = !1, c = !0;
  const e = {
    delta: 0,
    timestamp: 0,
    isProcessing: !1
  }, d = () => t = !0, n = a.reduce((s, r) => (s[r] = E(d), s), {}), { read: u, resolveKeyframes: i, update: f, preRender: m, render: R, postRender: g } = n, p = () => {
    const s = performance.now();
    t = !1, e.delta = c ? 1e3 / 60 : Math.max(Math.min(s - e.timestamp, F), 1), e.timestamp = s, e.isProcessing = !0, u.process(e), i.process(e), f.process(e), m.process(e), R.process(e), g.process(e), e.isProcessing = !1, t && l && (c = !1, o(p));
  }, h = () => {
    t = !0, c = !0, e.isProcessing || o(p);
  };
  return { schedule: a.reduce((s, r) => {
    const x = n[r];
    return s[r] = (P, v = !1, w = !1) => (t || h(), x.schedule(P, v, w)), s;
  }, {}), cancel: (s) => {
    for (let r = 0; r < a.length; r++)
      n[a[r]].cancel(s);
  }, state: e, steps: n };
}
export {
  B as createRenderBatcher,
  a as stepsOrder
};
