// @table-library/react-table-library@4.1.15/fromTreeToList-272db833.js
import { t as m } from "./toConsumableArray-25e5c43c.js";
import { d as v } from "./defineProperty-9f9de5d0.js";
var O = function(t) {
  var r;
  return !(t == null || (r = t.nodes) === null || r === void 0 || !r.length);
}, y = function t(r, e) {
  return r.reduce((function(n, o) {
    return n || (o.id === e ? o : o.nodes ? t(o.nodes, e) : n);
  }), null);
};
function p(t, r) {
  var e = Object.keys(t);
  if (Object.getOwnPropertySymbols) {
    var n = Object.getOwnPropertySymbols(t);
    r && (n = n.filter((function(o) {
      return Object.getOwnPropertyDescriptor(t, o).enumerable;
    }))), e.push.apply(e, n);
  }
  return e;
}
function s(t) {
  for (var r = 1; r < arguments.length; r++) {
    var e = arguments[r] != null ? arguments[r] : {};
    r % 2 ? p(Object(e), !0).forEach((function(n) {
      v(t, n, e[n]);
    })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(t, Object.getOwnPropertyDescriptors(e)) : p(Object(e)).forEach((function(n) {
      Object.defineProperty(t, n, Object.getOwnPropertyDescriptor(e, n));
    }));
  }
  return t;
}
var h = function t(r) {
  return (r || []).reduce((function(e, n) {
    return e = e.concat(n), n.nodes && (e = e.concat(t(n.nodes))), e;
  }), []);
}, P = function t(r, e, n) {
  var o = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : 0, f = arguments.length > 4 && arguments[4] !== void 0 ? arguments[4] : 0, c = arguments.length > 5 ? arguments[5] : void 0;
  return (e || []).reduce((function(a, u) {
    var l, i;
    i = u.nodes ? s(s({}, u), {}, { nodes: u.nodes.map((function(d) {
      return d.id;
    })) }) : u;
    var g = { treeXLevel: o, treeYLevel: f, parentNode: c || r, ancestors: c ? [c].concat(m((l = c == null ? void 0 : c.ancestors) !== null && l !== void 0 ? l : [])) : [c || r] };
    return i = s(s({}, i), g), a = a.concat(i), O(u) && n.includes(u.id) && (a = a.concat(t(r, u.nodes, n, o + 1, f, s(s({}, u), g)))), a;
  }), []);
};
export {
  h as a,
  P as b,
  y as f,
  O as h
};
