// @table-library/react-table-library@4.1.15/select.js
import { a as l, S as g } from "./select-d972db04.js";
import { o as C, _ as y } from "./styles-492c6342.js";
import * as p from "react";
import k from "../../../../clsx@1.1.1/node_modules/clsx/dist/clsx.m.js";
import { createElement as u } from "../../../../@emotion_react@11.14.0_@types_react@18.3.28_react@19.2.4/node_modules/@emotion/react/dist/emotion-react.browser.esm.js";
import { d as S, b as I, u as T } from "./Feature-0aee93bd.js";
import { C as j, i as O } from "./Cell-a4350b14.js";
import { a as w } from "./modifiers-d418f2b2.js";
import { H as P } from "./HeaderCell-789c2414.js";
import { s as B } from "./slicedToArray-c92cae3a.js";
import { d as N } from "./defineProperty-9f9de5d0.js";
import { R } from "./colors-953946b8.js";
import { u as x } from "./index-7b379f2c.js";
import { b as E } from "./useSyncControlledState-6e39bfdc.js";
var J = process.env.NODE_ENV === "production" ? { name: "e0dnmk", styles: "cursor:pointer" } : { name: "e0dnmk", styles: "cursor:pointer", map: "/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInN0eWxlcy50c3giXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQ21CIiwiZmlsZSI6InN0eWxlcy50c3giLCJzb3VyY2VzQ29udGVudCI6WyJpbXBvcnQgeyBjc3MgfSBmcm9tICdAZW1vdGlvbi9yZWFjdCc7XHJcbmV4cG9ydCBkZWZhdWx0IGNzcyBgXG4gIGN1cnNvcjogcG9pbnRlcjtcbmA7XHJcbiJdfQ== */", toString: function() {
  return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop).";
} }, W = p.forwardRef((function(o, t) {
  return u("input", y({ type: "checkbox", ref: t }, o, { css: J }));
})), h = function(o) {
  var t = o.checked, n = o.isIndeterminate, e = o.onChange;
  return u(W, { ref: function(r) {
    r && (t ? (r.indeterminate = !1, r.checked = !0) : n ? (r.indeterminate = !0, r.checked = !1) : (r.indeterminate = !1, r.checked = !1));
  }, type: "checkbox", onChange: e });
}, A = ["item"], M = p.memo((function(o) {
  var t = o.item, n = C(o, A), e = S(), r = I(), a = T();
  if (!e) throw new Error("No Select Context. No return value from useRowSelect provided to Table component.");
  var i = e.options.buttonSelect === l.SingleSelect && e.state.id === t.id || e.state.ids.includes(t.id), c = p.useCallback((function() {
    var s = e.options.buttonSelect === l.MultiSelect;
    a && s ? e.fns.onToggleByIdShift(t.id, e.options, w(r)) : s ? e.fns.onToggleByIdRecursively(t.id, { isCarryForward: e.options.isCarryForward, isPartialToAll: e.options.isPartialToAll }) : e.fns.onToggleByIdExclusively(t.id);
  }), [a, r, t.id, e]);
  return u(j, y({ stiff: !0 }, n), u(h, { checked: !!i, onChange: c }));
})), X = p.memo((function(o) {
  var t = S();
  if (!t) throw new Error("No Select Context. No return value from useRowSelect provided to Table component.");
  var n = t.state.all, e = !t.state.all && !t.state.none || t.options.buttonSelect === l.SingleSelect && t.state.id != null;
  return u(P, y({ stiff: !0 }, o), u(h, { checked: !!n, isIndeterminate: e, onChange: function() {
    return t.fns.onToggleAll({ isPartialToAll: t.options.isPartialToAll });
  } }));
}));
function b(o, t) {
  var n = Object.keys(o);
  if (Object.getOwnPropertySymbols) {
    var e = Object.getOwnPropertySymbols(o);
    t && (e = e.filter((function(r) {
      return Object.getOwnPropertyDescriptor(o, r).enumerable;
    }))), n.push.apply(n, e);
  }
  return n;
}
function f(o) {
  for (var t = 1; t < arguments.length; t++) {
    var n = arguments[t] != null ? arguments[t] : {};
    t % 2 ? b(Object(n), !0).forEach((function(e) {
      N(o, e, n[e]);
    })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(o, Object.getOwnPropertyDescriptors(n)) : b(Object(n)).forEach((function(e) {
      Object.defineProperty(o, e, Object.getOwnPropertyDescriptor(n, e));
    }));
  }
  return o;
}
var Z = function(o, t) {
  var n = o.item, e = t.select;
  if (!e) throw new Error("No 'select' in getRowProps. That's odd");
  var r = e.state.ids.includes(n.id), a = e.state.id === n.id;
  return { theme: `
    &.row-select-selected,
    &.row-select-single-selected {
      font-weight: bold;

      background-color: `.concat(R, `;
    }

    &.row-select-clickable {
      cursor: pointer;
    }
  `), className: k("row-select", { "row-select-clickable": e.options.clickType === g.RowClick, "row-select-selected": r, "row-select-single-selected": a }), onClick: function(i, c) {
    if (O(c) && e.options.clickType === g.RowClick) {
      var s = e.options.rowSelect === l.MultiSelect || e.options.buttonSelect === l.MultiSelect, d = e.options.rowSelect === l.MultiSelect, m = !!c.metaKey, v = !!c.shiftKey;
      m && s ? e.fns.onToggleById(i.id) : v && s ? e.fns.onToggleByIdShift(i.id, e.options, w(t)) : d ? e.fns.onToggleById(i.id) : e.fns.onToggleByIdExclusively(i.id);
    }
  } };
}, z = { ids: [], id: null }, G = { clickType: g.RowClick, rowSelect: l.SingleSelect, buttonSelect: l.MultiSelect, isCarryForward: !0, isPartialToAll: !1 }, te = function(o, t, n, e) {
  var r, a = f(f({}, z), (r = t == null ? void 0 : t.state) !== null && r !== void 0 ? r : {}), i = t != null && t.onChange ? t.onChange : function() {
  }, c = x(o, a, i, e), s = B(c, 2), d = s[0], m = s[1];
  return E("select", e, d), { state: d, fns: m, options: f(f({}, G), n ?? {}), _getRowProps: Z, components: { HeaderCellSelect: X, CellSelect: M } };
};
export {
  M as CellSelect,
  h as Checkbox,
  X as HeaderCellSelect,
  g as SelectClickTypes,
  l as SelectTypes,
  te as useRowSelect
};
