// @table-library/react-table-library@4.1.15/sort.js
import { S as E, a as D } from "./sort-c8713016.js";
import { o as N } from "./styles-492c6342.js";
import * as m from "react";
import { u as W, H as B } from "./HeaderCell-789c2414.js";
import R from "../../../../clsx@1.1.1/node_modules/clsx/dist/clsx.m.js";
import { createElement as i } from "../../../../@emotion_react@11.14.0_@types_react@18.3.28_react@19.2.4/node_modules/@emotion/react/dist/emotion-react.browser.esm.js";
import { d as I } from "./defineProperty-9f9de5d0.js";
import { B as $, I as F } from "./IconChevronSingleDown-c0866f17.js";
import { s as H } from "./slicedToArray-c92cae3a.js";
import { u as z, a as A, b as G } from "./useSyncControlledState-6e39bfdc.js";
function K(e, o) {
  var r = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var t = Object.getOwnPropertySymbols(e);
    o && (t = t.filter((function(n) {
      return Object.getOwnPropertyDescriptor(e, n).enumerable;
    }))), r.push.apply(r, t);
  }
  return r;
}
function b(e) {
  for (var o = 1; o < arguments.length; o++) {
    var r = arguments[o] != null ? arguments[o] : {};
    o % 2 ? K(Object(r), !0).forEach((function(t) {
      I(e, t, r[t]);
    })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(r)) : K(Object(r)).forEach((function(t) {
      Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(r, t));
    }));
  }
  return e;
}
var L = function(e) {
  var o = e.sort, r = e.sortKey, t = e.sortIcon, n = t === void 0 ? {} : t, a = e.children, u = o.state, P = o.fns, c = b(b({}, o.options.sortIcon), n), p = c.position === E.Prefix, h = c.position === E.Suffix, f = (function(g, j, v, w, l, O) {
    var y = { height: "".concat(v), width: "".concat(v) };
    return g.sortKey === j && g.reverse ? O ? m.cloneElement(O, b({}, y)) : null : g.sortKey !== j || g.reverse ? w ? m.cloneElement(w, b({}, y)) : null : l ? m.cloneElement(l, b({}, y)) : null;
  })(u, r, c.size, c.iconDefault, c.iconUp, c.iconDown);
  return i($, { className: R({ active: u.sortKey === r }), onClick: function() {
    return P.onToggleSort({ sortKey: r });
  } }, p && f && i("span", { style: { marginRight: c.margin } }, f), i("div", { title: typeof a == "string" ? a : "" }, a), h && f && i("span", { style: { marginLeft: c.margin } }, f));
}, U = ["sortKey", "sortIcon", "children"], M = m.memo((function(e) {
  var o = e.sortKey, r = e.sortIcon, t = r === void 0 ? {} : r, n = e.children, a = N(e, U), u = W();
  if (!u) throw new Error("No Sort Context. No return value from useSort provided to Table component.");
  return i(B, a, i(L, { sort: u, sortKey: o, sortIcon: t }, n));
}));
function T(e, o) {
  var r = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var t = Object.getOwnPropertySymbols(e);
    o && (t = t.filter((function(n) {
      return Object.getOwnPropertyDescriptor(e, n).enumerable;
    }))), r.push.apply(r, t);
  }
  return r;
}
function s(e) {
  for (var o = 1; o < arguments.length; o++) {
    var r = arguments[o] != null ? arguments[o] : {};
    o % 2 ? T(Object(r), !0).forEach((function(t) {
      I(e, t, r[t]);
    })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(r)) : T(Object(r)).forEach((function(t) {
      Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(r, t));
    }));
  }
  return e;
}
var k = "TOGGLE_SORT", _ = function(e, o) {
  switch (o.type) {
    case k:
      return (function(r, t) {
        var n = t.payload.value.sortKey === r.sortKey, a = r.reverse;
        if (n && a && t.payload.options.sortToggleType === D.AlternateWithReset) return { sortKey: "NONE", reverse: !1 };
        var u = n && !a;
        return s(s({}, t.payload.value), {}, { reverse: u });
      })(e, o);
    case "SET":
      return (function(r, t) {
        return s(s({}, r), t.payload);
      })(e, o);
    default:
      throw new Error();
  }
}, q = { sortKey: "NONE", reverse: !1 }, C = { position: E.Suffix, margin: "4px", size: "14px", iconDefault: i((function(e) {
  var o = e.width, r = e.height, t = e.viewBox, n = e.strokeWidth;
  return i("svg", { id: "svg-icon-chevron-single-up-down", "data-name": "svg-icon-chevron-single-up-down", "data-testid": "svg-icon-chevron-single-up-down", xmlns: "http://www.w3.org/2000/svg", width: o || "36rem", height: r || "36rem", viewBox: t || "0 0 36 36", strokeWidth: n || "0rem" }, i("polygon", { points: "36 12 36 15 18 6 0 15 0 12 18 3 36 12" }), i("polygon", { points: "0 24 0 21 18 30 36 21 36 24 18 33 0 24" }));
}), null), iconUp: i((function(e) {
  var o = e.width, r = e.height, t = e.viewBox, n = e.strokeWidth;
  return i("svg", { id: "svg-icon-chevron-single-up", "data-name": "svg-icon-chevron-single-up", "data-testid": "svg-icon-chevron-single-up", xmlns: "http://www.w3.org/2000/svg", width: o || "36rem", height: r || "36rem", viewBox: t || "0 0 36 36", strokeWidth: n || "0rem" }, i("polygon", { points: "36 21 36 24 18 15 0 24 0 21 18 12 36 21" }));
}), null), iconDown: i(F, null) }, J = { isServer: !1, sortToggleType: D.Alternate, sortIcon: C, isRecursive: !0 }, ne = function(e, o, r, t) {
  var n, a = s(s({}, q), (n = o == null ? void 0 : o.state) !== null && n !== void 0 ? n : {}), u = o != null && o.onChange ? o.onChange : function() {
  }, P = z(_, a, [], [u], t), c = H(P, 2), p = c[0], h = c[1], f = m.useCallback((function(l) {
    return h({ type: k, payload: { value: l, options: r } });
  }), [r, h]);
  A(a, p, (function() {
    return h({ type: "SET", payload: a });
  }));
  var g = m.useCallback((function(l, O, y) {
    var x = O[p.sortKey] || function(d) {
      return d;
    };
    return (p.reverse ? function(d) {
      return x(d).reverse();
    } : x)(l).reduce((function(d, S) {
      return y && S.nodes ? d.concat(s(s({}, S), {}, { nodes: g(S.nodes, O, y) })) : d.concat(S);
    }), []);
  }), [p]), j = m.useMemo((function() {
    return { onToggleSort: f };
  }), [f]);
  G("sort", t, p);
  var v = s(s(s({}, J), r), {}, { sortIcon: s(s({}, C), r ? r.sortIcon : {}) }), w = s(s({}, p), {}, { sortFn: g });
  return { state: w, fns: j, options: v, modifier: function(l) {
    return v.isServer ? l : w.sortFn(l, v.sortFns, v.isRecursive);
  }, components: { HeaderCellSort: M } };
};
export {
  M as HeaderCellSort,
  E as SortIconPositions,
  D as SortToggleType,
  ne as useSort
};
